### PR TITLE
Added Some Loss functions with some doc improvements

### PR DESCRIPTION
**Pull request by AdarshKumar712**
Added the following loss functions with tests:
1. mae
2. mean squared logarithmic error
3. huber loss
4. squared hinge loss
5. dice coeff loss
6. tversky loss 

Also added some documentation improvements for few other functions.

### CODE REVIEW

**Comment by AdarshKumar712** [comment]
@MikeInnes  @CarloLucibello  Can you please review these functions?

File: src/layers/stateless.jl [hunk 1]
```diff
@@ -2,9 +2,93 @@ using CuArrays
...
+  julia> mean_squared_logarithmic_error(y,y_)
+  3.771271382334686
+  ```
+Alias:
+  msle(ŷ,y;ϵ1=eps.(Float64.(ŷ)),ϵ2=eps.(Float64.(y)))
+
+"""
+mean_squared_logarithmic_error(ŷ, y;ϵ1=eps.(ŷ),ϵ2=eps.(eltype(ŷ).(y))) = sum((log.(ŷ+ϵ1).-log.(y+ϵ2)).^2) * 1 // length(y)
+#Alias
+msle(ŷ, y;ϵ1=eps.(ŷ),ϵ2=eps.(eltype(ŷ).(y))) = sum((log.(ŷ+ϵ1).-log.(y+ϵ2)).^2) * 1 // length(y)
```
Comment (dhairyagandhi96) [inline src/layers/stateless.jl]: Doesn't need a redefinition, maybe just use `const`

**Comment by CarloLucibello** [comment]
why you closed https://github.com/FluxML/Flux.jl/pull/1019? btw, I left some pending comments over there

**Comment by AdarshKumar712** [comment]
Actually that PR had been static since quite long, also I had to make some additional changes to that. So I thought better close that one. Really sorry for the inconvenience caused to you. I was unaware of that. Is there anything I can do now about that?

File: src/layers/stateless.jl (hunk 1 above)
Comment (AdarshKumar712) [reply src/layers/stateless.jl]: Ok, I will replace that with const.

Comment (dhairyagandhi96) [reply src/layers/stateless.jl]: Actually, now that I look at it, we don't need two names for something that does the same thing, maybe remove the long form definition here

Comment (AdarshKumar712) [reply src/layers/stateless.jl]: Actually, I added that long form as msle isn't that general term, but the full form seems to explain pretty much about the loss function. However, if you say, I will just remove the long form.

**Comment by AdarshKumar712** [comment]
@dhairyagandhi96 Are there any other changes, that I should do here?

File: src/layers/stateless.jl [hunk 2]
```diff
@@ -2,9 +2,88 @@ using CuArrays
...
+  hub_loss = dtype(0)
+  for i in 1:length(y)
+    if (abs_error[i]<=delta)
+      hub_loss+=abs_error[i]^2*dtype(0.5)
+    else
+      hub_loss+=delta*(abs_error[i]- dtype(0.5*delta))
+    end
+  
+  return hub_loss*1//length(y)
+  end
```
Comment (lungd) [inline src/layers/stateless.jl]: The return statement should not be inside the inner loop?
And I think that Julia ppl prefer to omit the 'return' keyword
```suggestion
  end
  hub_loss*1//length(y)
```

File: test/layers/stateless.jl [hunk 3]
```diff
@@ -12,6 +12,20 @@ const ϵ = 1e-7
   @testset "mse" begin
     @test mse(ŷ, y) ≈ (.1^2 + .9^2)/2
   end
+  
+  @testset "mae" begin
+    @test Flux.mae(ŷ, y) ≈ 1/2
+  end
+  
+  @testset "huber_loss" begin
+    @test Flux.huber_loss(ŷ, y) ≈ 0.0012499999999999994
```
Comment (lungd) [inline test/layers/stateless.jl]: With my suggestion from above:

```suggestion
    @test Flux.huber_loss(ŷ, y) ≈ 0.20500000000000002
```

File: src/layers/stateless.jl (hunk 2 above)
Comment (AdarshKumar712) [reply src/layers/stateless.jl]: Oh yes!, that should have been outside the for loop. Thanks for pointing.

**Comment by AdarshKumar712** [comment]
@CarloLucibello can you please suggest here those changes you had in pending comments in previous PR?

File: src/layers/stateless.jl [hunk 4]
```diff
@@ -2,9 +2,88 @@ using CuArrays
 using NNlib: logsoftmax, logσ
 
 # Cost functions
+"""
+    mae(ŷ, y)
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: newline below

File: src/layers/stateless.jl [hunk 5]
```diff
@@ -2,9 +2,88 @@ using CuArrays
...
+L1 loss function. Computes the mean of absolute error between prediction and true values
+"""
+mae(ŷ, y) = sum(abs.(ŷ .- y)) * 1 // length(y)
+
 
+"""
+    mse(ŷ, y)
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: new line below

File: src/layers/stateless.jl [hunk 6]
```diff
@@ -2,9 +2,88 @@ using CuArrays
...
+
+  julia> ŷ = [12466.1,16353.95,16367.98]
+  3-element Array{Float64,1}:
+  12466.1 
+  16353.95
+  16367.98
+
+  julia> msle(ŷ,y)
+  3.771271382334686
+  ```
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: numerical examples like this are useless. Better put the explicit mathematical definition in the doscstring

File: src/layers/stateless.jl [hunk 7]
```diff
@@ -2,9 +2,88 @@ using CuArrays
...
+    mse(ŷ, y)
+L2 loss function. Computes the mean of the squared errors between prediction and true values
+"""
 mse(ŷ, y) = sum((ŷ .- y).^2) * 1 // length(y)
 
+
+"""
+    msle(ŷ, y;ϵ1=eps.(Float64.(ŷ)),ϵ2=eps.(Float64.(y)))
+
+Mean Squared Logarithmic Error,an L2 loss function. Returns the mean of the squared logarithmic errors of prediction ŷ, and true values y. The ϵ1 and ϵ2 terms provide numerical stability.
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: please remove "an L2 loss function", doesn't make much sense

File: src/layers/stateless.jl [hunk 8]
```diff
@@ -2,9 +2,88 @@ using CuArrays
...
+  1.73  
+
+  julia> y_= [-1.376,0,3.37]
+  3-element Array{Float64,1}:
+  -1.376
+   0.0  
+   3.37 
+
+  julia> huber_loss(y,y_)
+  0.7131999999999998
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: same here , remove numerical example, give mathematical definition

**Comment by CarloLucibello** [comment]
needs a rebase

File: src/layers/stateless.jl (hunk 4 above)
Comment (AdarshKumar712) [reply src/layers/stateless.jl]: done!

**Comment by AdarshKumar712** [comment]
@CarloLucibello done!

File: src/layers/stateless.jl [hunk 9]
```diff
@@ -7,6 +15,41 @@ Return the mean squared error `sum((ŷ .- y).^2) / length(y)`.
 mse(ŷ, y) = sum((ŷ .- y).^2) * 1 // length(y)
 
 
+"""
+    msle(ŷ, y;ϵ1=eps.(Float64.(ŷ)),ϵ2=eps.(Float64.(y)))
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: maybe just single epsilone jere? also add whitespace after `;` and `,`

File: src/layers/stateless.jl [hunk 10]
```diff
@@ -7,6 +15,41 @@ Return the mean squared error `sum((ŷ .- y).^2) / length(y)`.
...
+The ϵ1 and ϵ2 terms provide numerical stability. This error penalizes an under-predicted estimate greater than an over-predicted estimate.
+"""
+msle(ŷ, y;ϵ1=eps.(ŷ),ϵ2=eps.(eltype(ŷ).(y))) = sum((log.(ŷ+ϵ1).-log.(y+ϵ2)).^2) * 1 // length(y)
+
+
+
+"""
+    huber_loss(ŷ, y,delta=1.0)
+
+Computes the mean of the Huber loss. By default, delta is set to 1.0.
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: newline below

File: src/layers/stateless.jl [hunk 11]
```diff
@@ -7,6 +15,41 @@ Return the mean squared error `sum((ŷ .- y).^2) / length(y)`.
...
+  dtype= eltype(ŷ)
+  delta = dtype(delta)
+  hub_loss = dtype(0)
+  for i in 1:length(y)
+    if (abs_error[i]<=delta)
+      hub_loss+=abs_error[i]^2*dtype(0.5)
+    else
+      hub_loss+=delta*(abs_error[i]- dtype(0.5*delta))
+    end
+  end
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: this is not gpu friendly, should use `sum` and broadcasting

File: src/layers/stateless.jl [hunk 12]
```diff
@@ -122,6 +167,44 @@ poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
...
+See also [`squared_hinge`](@ref).
 """
 hinge(ŷ, y) = sum(max.(0, 1 .-  ŷ .* y)) *1 // size(y,2)
+
+"""
+    squared_hinge(ŷ, y)
+
+Computes squared hinge loss given the prediction `ŷ` and true labels `y` (conatining 1 or -1)
+
+See also [`hinge`](@ref).
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: this could use a mathematical definition

File: src/layers/stateless.jl [hunk 13]
```diff
@@ -122,6 +167,44 @@ poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
...
+function dice_coeff_loss(y_pred,y_true,smooth=eltype(y_pred)(1.0))
+    intersection = sum(y_true.*y_pred)
+    return 1 - (2*intersection + smooth)/(sum(y_true.^2) + sum(y_pred.^2)+smooth)
+end
+
+"""
+    tversky_loss(y_pred,y_true,beta = 0.7)
+
+Used with imbalanced data to give more weightage to False negatives. Larger β weigh recall higher than precision (by placing more emphasis on false negatives)
+    tversky_loss(ŷ,y,beta) = 1 - sum(|y.*ŷ| + 1) / (sum(y.*ŷ + beta*(1 .- y).*ŷ + (1 .- beta)*y.*(1 .- ŷ))+ 1)
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: newline

**Comment by CarloLucibello** [comment]
good, we are almost there. All those functions should be referenced in the docs

File: src/layers/stateless.jl (hunk 11 above)
Comment (AdarshKumar712) [reply src/layers/stateless.jl]: I have changed the function. Hopefully, now it's gpu friendly

File: src/layers/stateless.jl [hunk 14]
```diff
@@ -7,6 +15,34 @@ Return the mean squared error `sum((ŷ .- y).^2) / length(y)`.
 mse(ŷ, y) = sum((ŷ .- y).^2) * 1 // length(y)
 
 
+"""
+    msle(ŷ, y; ϵ1=eps.(Float64.(ŷ)))
+
+Mean Squared Logarithmic Error. Returns the mean of the squared logarithmic errors `sum((log.(ŷ+ϵ1) .- log.(y+ϵ2)).^2) * 1 / length(y)`.<br>
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: just ϵ

File: src/layers/stateless.jl [hunk 15]
```diff
@@ -122,6 +160,48 @@ poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
     hinge(ŷ, y)
 
 Measures the loss given the prediction `ŷ` and true labels `y` (containing 1 or -1). 
-[Hinge Loss](https://en.wikipedia.org/wiki/Hinge_loss).
+Returns `sum((max.(0,1 .-ŷ .* y))) *1 // size(y, 2)`
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: let's avoid implementation detail as *1 here. Also, whitespace after comma and .-

File: src/layers/stateless.jl [hunk 16]
```diff
@@ -122,6 +160,48 @@ poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
...
+[Hinge Loss](https://en.wikipedia.org/wiki/Hinge_loss)
+See also [`squared_hinge`](@ref).
 """
 hinge(ŷ, y) = sum(max.(0, 1 .-  ŷ .* y)) *1 // size(y,2)
+
+"""
+    squared_hinge(ŷ, y)
+
+Computes squared hinge loss given the prediction `ŷ` and true labels `y` (conatining 1 or -1).
+Returns `sum((max.(0,1 .-ŷ .* y)).^2) *1 // size(y, 2)`
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: same

File: src/layers/stateless.jl [hunk 17]
```diff
@@ -122,6 +160,48 @@ poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
...
+Loss function used in Image Segmentation. Calculates loss based on dice coefficient. Similar to F1_score.
+    
+    Dice_Coefficient(A,B) = 2 * sum( |A*B| + smooth) / (sum( A^2 ) + sum( B^2 )+ smooth)
+    Dice_loss = 1 - Dice_Coefficient
+
+Ref: [V-Net: Fully Convolutional Neural Networks forVolumetric Medical Image Segmentation](https://arxiv.org/pdf/1606.04797v1.pdf)
+"""
+function dice_coeff_loss(y_pred, y_true; smooth=eltype(y_pred)(1.0))
+    intersection = sum(y_true.*y_pred)
+    return 1 - (2*intersection + smooth)/(sum(y_true.^2) + sum(y_pred.^2)+smooth)
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: y\hat and y for consistency with other losses

File: src/layers/stateless.jl [hunk 18]
```diff
@@ -122,6 +160,48 @@ poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
...
+end
+
+"""
+    tversky_loss(y_pred, y_true, beta = 0.7)
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: let's use unicode beta here and in code

File: src/layers/stateless.jl [hunk 19]
```diff
@@ -122,6 +160,48 @@ poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
...
+
+Used with imbalanced data to give more weightage to False negatives. Larger β weigh recall higher than precision (by placing more emphasis on false negatives)
+    
+    tversky_loss(ŷ,y,beta) = 1 - sum(|y.*ŷ| + 1) / (sum(y.*ŷ + beta*(1 .- y).*ŷ + (1 .- beta)*y.*(1 .- ŷ))+ 1)
+
+Ref: [Tversky loss function for image segmentation using 3D fully convolutional deep networks](https://arxiv.org/pdf/1706.05721.pdf)
+"""
+function tversky_loss(y_pred, y_true; beta = eltype(y_pred)(0.7))
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: same as above

File: test/layers/stateless.jl [hunk 20]
```diff
@@ -64,18 +78,36 @@ const ϵ = 1e-7
...
+  y1 = [0 1.4 0.5 1.2]
+  @testset "dice_coeff_loss" begin
+    @test Flux.dice_coeff_loss(y, y1) ≈ 0.2799999999999999
+    @test Flux.dice_coeff_loss(y,y) ≈ 0.0
+  end
+            
+  @testset "tversky_loss" begin
+    @test Flux.tversky_loss(y,y1) ≈ 0.028747433264887046
+    @test Flux.tversky_loss(y,y1,beta = 0.8) ≈ 0.050200803212851364
+    @test Flux.tversky_loss(y,y) ≈ -0.5576923076923075
```
Comment (CarloLucibello) [inline test/layers/stateless.jl]: check in all tests that argument order is correct: it should be ŷ first, y second

File: test/layers/stateless.jl [hunk 21]
```diff
@@ -64,18 +78,36 @@ const ϵ = 1e-7
...
+  end
+            
   @testset "no spurious promotions" begin
     for T in (Float32, Float64)
       y = rand(T, 2)
       ŷ = rand(T, 2)
-      for f in (mse, crossentropy, logitcrossentropy, Flux.kldivergence, Flux.hinge, Flux.poisson)
+      for f in (mse, crossentropy, logitcrossentropy, Flux.kldivergence, Flux.hinge, Flux.poisson,Flux.mae,Flux.huber_loss,Flux.msle,Flux.squared_hinge,Flux.dice_coeff_loss,Flux.tversky_loss)
```
Comment (CarloLucibello) [inline test/layers/stateless.jl]: line too long, use newline

File: src/layers/stateless.jl [hunk 22]
```diff
@@ -122,6 +160,48 @@ poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
...
+"""
+function dice_coeff_loss(y_pred, y_true; smooth=eltype(y_pred)(1.0))
+    intersection = sum(y_true.*y_pred)
+    return 1 - (2*intersection + smooth)/(sum(y_true.^2) + sum(y_pred.^2)+smooth)
+end
+
+"""
+    tversky_loss(y_pred, y_true, beta = 0.7)
+
+Used with imbalanced data to give more weightage to False negatives. Larger β weigh recall higher than precision (by placing more emphasis on false negatives)
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: here and in some other places the line is too long

File: src/layers/stateless.jl (hunk 15 above)
Comment (AdarshKumar712) [reply src/layers/stateless.jl]: Should I write it as `sum((max.(0, 1 .- ŷ .* y))) / size(y, 2)` ?

File: src/layers/stateless.jl (hunk 22 above)
Comment (AdarshKumar712) [reply src/layers/stateless.jl]: Ok I will break the lines

File: test/layers/stateless.jl (hunk 20 above)
Comment (AdarshKumar712) [reply test/layers/stateless.jl]: Can you please explain this a bit? I didn't get it.

File: src/layers/stateless.jl (hunk 18 above)
Comment (AdarshKumar712) [reply src/layers/stateless.jl]: Ok will change that

File: test/layers/stateless.jl (hunk 20 above)
Comment (AdarshKumar712) [reply test/layers/stateless.jl]: You mean to check for loss(y1,y) also?

**Comment by AdarshKumar712** [comment]
@CarloLucibello I have made the changes asked to do. Please have a look.

File: src/layers/stateless.jl [hunk 23]
```diff
@@ -122,6 +163,49 @@ poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
...
 """
-hinge(ŷ, y) = sum(max.(0, 1 .-  ŷ .* y)) *1 // size(y,2)
+squared_hinge(ŷ, y) = sum((max.(0, 1 .- ŷ .* y)).^2) *1 // size(y, 2)
+
+"""
+    dice_coeff_loss(ŷ, y, smooth = 1)
+
+Loss function used in Image Segmentation. Calculates loss based on dice coefficient. Similar to F1_score.
+    
+    Dice_Coefficient(ŷ, y) = 2 * sum( |ŷ.* y| + smooth) / (sum( ŷ.^2 ) + sum( y.^2 ) + smooth)
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]:
```suggestion
    Dice_Coefficient(ŷ, y) = 2 * sum(|ŷ .* y| + smooth) / (sum(ŷ.^2) + sum(y.^2) + smooth)
```

File: src/layers/stateless.jl [hunk 24]
```diff
@@ -7,6 +15,36 @@ Return the mean squared error `sum((ŷ .- y).^2) / length(y)`.
 mse(ŷ, y) = sum((ŷ .- y).^2) * 1 // length(y)
 
 
+"""
+    msle(ŷ, y; ϵ = eps.(Float64.(ŷ)))
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: eps.(Float64.(ŷ)) ???

File: src/layers/stateless.jl [hunk 25]
```diff
@@ -7,6 +15,36 @@ Return the mean squared error `sum((ŷ .- y).^2) / length(y)`.
...
+The `ϵ` term provides numerical stability. 
+
+This error penalizes an under-predicted estimate greater than an over-predicted estimate.
+"""
+msle(ŷ, y; ϵ = eps.(ŷ)) = sum((log.(ŷ + ϵ).-log.(y + ϵ)).^2) * 1 // length(y)
+
+
+
+"""
+    huber_loss(ŷ, y; delta = 1.0)
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: delta ->  δ

File: src/layers/stateless.jl [hunk 26]
```diff
@@ -122,6 +163,49 @@ poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
...
+Computes squared hinge loss given the prediction `ŷ` and true labels `y` (conatining 1 or -1).
+Returns `sum((max.(0, 1 .- ŷ .* y)).^2) / size(y, 2)`
+
+See also [`hinge`](@ref).
 """
-hinge(ŷ, y) = sum(max.(0, 1 .-  ŷ .* y)) *1 // size(y,2)
+squared_hinge(ŷ, y) = sum((max.(0, 1 .- ŷ .* y)).^2) *1 // size(y, 2)
+
+"""
+    dice_coeff_loss(ŷ, y, smooth = 1)
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: ,  -> ;

File: src/layers/stateless.jl [hunk 27]
```diff
@@ -122,6 +163,49 @@ poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
...
+
+Loss function used in Image Segmentation. Calculates loss based on dice coefficient. Similar to F1_score.
+    
+    Dice_Coefficient(ŷ, y) = 2 * sum( |ŷ.* y| + smooth) / (sum( ŷ.^2 ) + sum( y.^2 ) + smooth)
+    Dice_loss = 1 - Dice_Coefficient
+
+[V-Net: Fully Convolutional Neural Networks forVolumetric Medical Image Segmentation](https://arxiv.org/pdf/1606.04797v1.pdf)
+"""
+function dice_coeff_loss(ŷ, y; smooth = eltype(ŷ)(1.0))
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: for keyword arguments, avoid whitespace around =

File: src/layers/stateless.jl [hunk 28]
```diff
@@ -122,6 +163,49 @@ poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
...
+    intersection = sum(y.*ŷ)
+    return 1 - (2*intersection + smooth) / (sum(y.^2) + sum(ŷ.^2) + smooth)
+end
+
+"""
+    tversky_loss(ŷ, y, β = 0.7)
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: , -> ;

File: src/layers/stateless.jl [hunk 29]
```diff
@@ -7,6 +15,36 @@ Return the mean squared error `sum((ŷ .- y).^2) / length(y)`.
...
+
+                    | 0.5*|ŷ - y|,   for |ŷ - y| <= delta
+      Hubber loss = |
+                    | delta*(|ŷ- y| - 0.5*delta),  otherwise
+
+[`Huber Loss`](https://en.wikipedia.org/wiki/Huber_loss).
+"""
+function huber_loss(ŷ, y; delta = eltype(ŷ)(1))
+   abs_error = abs.(ŷ.-y)
+   temp = abs_error.<delta
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: whitespaces around operators

File: src/layers/stateless.jl [hunk 30]
```diff
@@ -7,6 +15,36 @@ Return the mean squared error `sum((ŷ .- y).^2) / length(y)`.
...
 
+"""
+    msle(ŷ, y; ϵ = eps.(Float64.(ŷ)))
+
+Returns the mean of the squared logarithmic errors `sum((log.(ŷ + ϵ) .- log.(y + ϵ)).^2) / length(y)`.
+The `ϵ` term provides numerical stability. 
+
+This error penalizes an under-predicted estimate greater than an over-predicted estimate.
+"""
+msle(ŷ, y; ϵ = eps.(ŷ)) = sum((log.(ŷ + ϵ).-log.(y + ϵ)).^2) * 1 // length(y)
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: ϵ = eps.(ŷ) is creating a whole array. Let's use eps(eltype(ŷ)).

**Comment by AdarshKumar712** [comment]
@CarloLucibello  I have made certain changes. I have changed the format of mathematical definitions of the tversky loss and dice_coeff_loss, made them similar to rest of the cost functions. Please have a look.

**Comment by CarloLucibello** [comment]
looks good, thanks!

bors r+

**Comment by AdarshKumar712** [comment]
Thanks @CarloLucibello ! :)

**Pull request merged**